### PR TITLE
feat: init policy template

### PR DESCRIPTION
**Pull request by ytvnr**

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+## Issue
+
+https://gravitee.atlassian.net/browse/APIM-XXX
```
Comment (guillaumelamirand) [inline .github/pull_request_template.md]: Not sure this PR template will apply on this kind of policy. wdyt?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+version: 2.1
+
+# this allows you to use CircleCI's dynamic configuration feature
+setup: true
+
+orbs:
+    gravitee: gravitee-io/gravitee@4.0.0
```
Comment (guillaumelamirand) [inline .circleci/config.yml]:
```suggestion
    gravitee: gravitee-io/gravitee@4.1.0
```

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        <artifactId>gravitee-parent</artifactId>
+        <version>21.0.1</version>
+    </parent>
+
+    <properties>
+        <gravitee-bom.version>5.0.0</gravitee-bom.version>
+        <gravitee-gateway-api.version>3.0.0-alpha.9</gravitee-gateway-api.version>
+        <gravitee-policy-api.version>1.11.0</gravitee-policy-api.version>
+        <gravitee-common.version>2.1.0</gravitee-common.version>
+        <gravitee-apim-gateway-tests-sdk.version>4.0.0-alpha.1</gravitee-apim-gateway-tests-sdk.version>
```
Comment (guillaumelamirand) [inline pom.xml]: As this property is not only used for the SDK, I would name it just `gravitee-apim.version`

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                <scope>import</scope>
+                <type>pom</type>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+
+    <dependencies>
+        <!-- Provided scope -->
+        <dependency>
+            <groupId>io.gravitee.gateway</groupId>
```
Comment (guillaumelamirand) [inline pom.xml:81]: You should put in int the dependencyManagement to force the version.

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            <artifactId>gravitee-gateway-api</artifactId>
+            <version>${gravitee-gateway-api.version}</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.gravitee.policy</groupId>
+            <artifactId>gravitee-policy-api</artifactId>
+            <version>${gravitee-policy-api.version}</version>
```
Comment (guillaumelamirand) [inline pom.xml]: Same here

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.gravitee.common</groupId>
+            <artifactId>gravitee-common</artifactId>
+            <version>${gravitee-common.version}</version>
```
Comment (guillaumelamirand) [inline pom.xml]: same here

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            <version>${gravitee-apim-gateway-tests-sdk.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.gravitee.apim.plugin.endpoint</groupId>
+            <artifactId>gravitee-apim-plugin-endpoint-http-proxy</artifactId>
+            <version>${gravitee-apim-gateway-tests-sdk.version}</version>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (guillaumelamirand) [inline pom.xml:128]: What about message policy ?

File: src/assembly/policy-assembly.xml [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            <includes>
+                <include>README.adoc</include>
+            </includes>
+            <outputDirectory>docs</outputDirectory>
+        </fileSet>
+
+        <fileSet>
+            <directory>${basedir}</directory>
+            <includes>
+                <include>${icon}</include>
```
Comment (guillaumelamirand) [inline src/assembly/policy-assembly.xml:52]: where the icon variable comes from ?

File: src/main/java/io/gravitee/policy/template/TemplatePolicy.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+@Slf4j
+public class TemplatePolicy implements Policy {
+
+    public static final String TEMPLATE_POLICY_HEADER = "x-template-policy";
+    public static final String TEMPLATE_POLICY_EXECUTED_HEADER = "x-template-policy-executed";
+    public static final String ERROR_MESSAGE = "Invalid header";
+
+    private final TemplatePolicyConfiguration configuration;
+
+    public TemplatePolicy(TemplatePolicyConfiguration configuration) {
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/policy/template/TemplatePolicy.java]: @RequiredArgs....

File: src/main/java/io/gravitee/policy/template/TemplatePolicy.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    @Override
+    public Completable onResponse(HttpExecutionContext ctx) {
+        if (shouldInterrupt(ctx.response().headers())) {
+            return ctx.interruptWith(new ExecutionFailure(HttpStatusCode.INTERNAL_SERVER_ERROR_500).message(ERROR_MESSAGE));
+        }
+        ctx.response().headers().add(TEMPLATE_POLICY_EXECUTED_HEADER, "ok");
+        return Policy.super.onRequest(ctx);
+    }
+
+    private boolean shouldInterrupt(HttpHeaders ctx) {
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/policy/template/TemplatePolicy.java]:
```suggestion
    private boolean shouldInterrupt(HttpHeaders httpHeaders) {
```

File: README.adoc [hunk 11]
```diff
@@ -17,7 +17,12 @@
 
 == Description
 
-Policy template
+A policy template to fork and use as a quick starter.
+
+This policy will compare `X-Template-Policy` header value with its configuration `errorKey` field, if both values are equal, then the policy will make the request or response fail.
```
Comment (guillaumelamirand) [inline README.adoc]:
```suggestion
This policy will compare `X-Template-Policy` header value with its configuration `errorKey` field, if both values are equal, then the policy will interrupt the request with a failure execution error.
```

File: src/main/java/io/gravitee/policy/template/TemplatePolicy.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            ctx.request().headers().add(TEMPLATE_POLICY_EXECUTED_HEADER, "ok");
+            return Completable.complete();
+        });
+    }
+
+    @Override
+    public Completable onResponse(HttpExecutionContext ctx) {
+        if (shouldInterrupt(ctx.response().headers())) {
+            return ctx.interruptWith(new ExecutionFailure(HttpStatusCode.INTERNAL_SERVER_ERROR_500).message(ERROR_MESSAGE));
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/policy/template/TemplatePolicy.java:75]: Why does the error code is different here ?

**Review by guillaumelamirand — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 7 above)
Comment (ytvnr) [reply pom.xml:128]: is the question related to a particular line of the pom ?
For the moment, I only did the "OSS part" of the policy template.
I have not a clear strategy about how to integrate event native related code / dependencies here. But that should definitely be discussed

File: src/assembly/policy-assembly.xml (hunk 8 above)
Comment (ytvnr) [reply src/assembly/policy-assembly.xml:52]: it's the icon field read from `plugin.properties` thanks to `properties-maven-plugin`

File: src/main/java/io/gravitee/policy/template/TemplatePolicy.java (hunk 12 above)
Comment (ytvnr) [reply src/main/java/io/gravitee/policy/template/TemplatePolicy.java:75]: as discussed, it depends on the current phase (would be weird to interrupt response with a 400)

**Review by gaetanmaisse — APPROVED** [review]

**Pull request merged**